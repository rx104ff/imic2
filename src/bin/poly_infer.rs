use std::env;
use imic2::{common, poly_infer}; 

fn main() {
    let args: Vec<String> = env::args().collect();
    if args.len() < 2 {
        eprintln!("Usage: cargo run --bin poly_infer \"<JUDGMENT>\"");
        eprintln!("Example: cargo run --bin poly_infer \"f: 'a.'a -> 'a |- f 3\"");
        return;
    }

    let input = &args[1];

    // 1. Tokenize the input string.
    let tokens = common::tokenizer::tokenize(input);

    // 2. Parse the tokens into a Judgment struct.
    let mut parser = poly_infer::parser::Parser::new(tokens);
    let (judgment, used_names) = match parser.parse_judgment() {
        Ok(j) => j,
        Err(e) => {
            eprintln!("Parsing Error: {}", e);
            return;
        }
    };

    // 3. Run the type inferrer on the parsed judgment.
    match poly_infer::poly_infer::infer_judgment(&judgment, used_names) {
        Ok(derivation) => {
            // The derivation object contains the full proof tree with all types resolved.
            // We can now print it directly.
            println!("{}", derivation);
        },
        Err(e) => eprintln!("Type Error: {}", e),
    }
}
