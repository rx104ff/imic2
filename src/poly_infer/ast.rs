use std::cell::Cell;
use std::fmt;
use std::collections::HashSet;
use std::sync::atomic::{AtomicUsize, Ordering};
use std::hash::{Hash, Hasher};

// --- Core Expression AST for PolyTypingML4 ---

#[derive(Debug, Clone, PartialEq, Eq, Hash)]
pub struct Var(pub String);

#[derive(Debug, Clone, PartialEq)]
pub enum Op { Add, Sub, Mul, Lt, Cons }

#[derive(Debug, Clone, PartialEq)]
pub enum Expr {
    Int(i64),
    Bool(bool),
    Var(Var),
    Let(Var, Box<Expr>, Box<Expr>, bool),
    LetRec(Var, Var, Box<Expr>, Box<Expr>, bool),
    Fun(Var, Box<Expr>, bool),
    App(Box<Expr>, Box<Expr>, bool),
    If(Box<Expr>, Box<Expr>, Box<Expr>, bool),
    BinOp(Box<Expr>, Op, Box<Expr>, bool),
    Nil,
    Match(Box<Expr>, Box<Expr>, Var, Var, Box<Expr>, bool),
}

// --- Types, Type Variables, and Type Schemes ---

thread_local! {
    static NEXT_TYPE_VAR_ID: Cell<usize> = Cell::new(0);
}

// NEW: Public function to reset the counter for deterministic tests.
pub fn reset_type_var_counter() {
    NEXT_TYPE_VAR_ID.with(|counter| {
        counter.set(0);
    });
}

impl TypeVar {
    /// Creates a fresh type variable for internal use by the inferrer.
    pub fn new() -> Self {
        let id = NEXT_TYPE_VAR_ID.with(|counter| {
            let id = counter.get();
            counter.set(id + 1);
            id
        });
        let name = format!("'{}", ((id % 26) as u8 + b'a') as char);
        TypeVar { id, name }
    }


    /// Creates a type variable from a name found in the source code.
    pub fn new_from_name(name: String) -> Self {
        let id = NEXT_TYPE_VAR_ID.with(|counter| {
            let id = counter.get();
            counter.set(id + 1);
            id
        });
        TypeVar { id, name: format!("'{}", name) }
    }
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub enum Type {
    Int,
    Bool,
    Fun(Box<Type>, Box<Type>),
    List(Box<Type>),
    Var(TypeVar), // Represents an unknown type, e.g., 'a
}

/// A Type Scheme represents a polymorphic type, e.g., `forall 'a. 'a -> 'a`.
#[derive(Debug, Clone)]
pub struct TypeVar {
    pub id: usize, // Made public for hashing/eq
    name: String,
}

#[derive(Debug, Clone)]
pub struct TyScheme {
    pub vars: Vec<TypeVar>,
    pub ty: Type,
}

impl PartialEq for TyScheme {
    fn eq(&self, other: &Self) -> bool {
        if self.ty != other.ty { return false; }
        let self_vars: HashSet<_> = self.vars.iter().collect();
        let other_vars: HashSet<_> = other.vars.iter().collect();
        self_vars == other_vars
    }
}
impl Eq for TyScheme {}

impl PartialEq for TypeVar { fn eq(&self, other: &Self) -> bool { self.id == other.id } }
impl Eq for TypeVar {}
impl Hash for TypeVar { fn hash<H: Hasher>(&self, state: &mut H) { self.id.hash(state); } }

/// The Type Environment maps variables to polymorphic Type Schemes.
pub type TypeEnv = Vec<(Var, TyScheme)>;

/// The Judgment represents the input to the type inferrer: `Γ ⊢ e`.
#[derive(Debug, Clone, PartialEq)]
pub struct Judgment {
    pub env: TypeEnv,
    pub expr: Expr,
    pub ty: Type,
}

// --- Display Implementations for Pretty-Printing ---

impl fmt::Display for Var {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(f, "{}", self.0)
    }
}

impl fmt::Display for Op {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        match self {
            Op::Add => write!(f, "+"), Op::Sub => write!(f, "-"), Op::Mul => write!(f, "*"),
            Op::Lt => write!(f, "<"), Op::Cons => write!(f, "::"),
        }
    }
}

impl fmt::Display for Expr {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        match self {
            Expr::Int(n) => write!(f, "{}", n),
            Expr::Bool(b) => write!(f, "{}", b),
            Expr::Var(v) => write!(f, "{}", v.0),
            Expr::Nil => write!(f, "[]"),
            Expr::Let(x, e1, e2, is_paren) => {
                let s = format!("let {} = {} in {}", x, e1, e2);
                if *is_paren { write!(f, "({})", s) } else { write!(f, "{}", s) }
            },
            Expr::LetRec(func, param, body, cont, is_paren) => {
                let s = format!("let rec {} = fun {} -> {} in {}", func, param, body, cont);
                if *is_paren { write!(f, "({})", s) } else { write!(f, "{}", s) }
            },
            Expr::If(c, t, e, is_paren) => {
                let s = format!("if {} then {} else {}", c, t, e);
                if *is_paren { write!(f, "({})", s) } else { write!(f, "{}", s) }
            },
            Expr::BinOp(e1, op, e2, is_paren) => {
                let s = format!("{} {} {}", e1, op, e2);
                if *is_paren { write!(f, "({})", s) } else { write!(f, "{}", s) }
            }
            Expr::Fun(p, b, is_paren) => {
                let s = format!("fun {} -> {}", p, b);
                if *is_paren { write!(f, "({})", s) } else { write!(f, "{}", s) }
            },
            Expr::App(e1, e2, is_paren) => {
                let s = format!("{} {}", e1, e2);
                if *is_paren { write!(f, "({})", s) } else { write!(f, "{}", s) }
            },
            Expr::Match(e, nil_case, x, y, cons_case, is_paren) => {
                let s = format!("match {} with [] -> {} | {}::{} -> {}", e, nil_case, x, y, cons_case);
                if *is_paren { write!(f, "({})", s) } else { write!(f, "{}", s) }
            },
        }
    }
}

impl fmt::Display for TypeVar {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(f, "{}", self.name)
    }
}

impl fmt::Display for Type {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        match self {
            Type::Int => write!(f, "int"),
            Type::Bool => write!(f, "bool"),
            Type::Fun(t1, t2) => write!(f, "({} -> {})", t1, t2),
            Type::List(t) => write!(f, "{} list", t),
            Type::Var(tv) => write!(f, "{}", tv),
        }
    }
}

impl fmt::Display for TyScheme {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        if self.vars.is_empty() {
            write!(f, "{}", self.ty)
        } else {
            let mut sorted_vars = self.vars.clone();
            // Sort by the string name to ensure 'a' comes before 'b'.
            sorted_vars.sort_by(|a, b| a.name.cmp(&b.name));

            let vars_str = sorted_vars.iter().map(|v| v.to_string()).collect::<Vec<_>>().join(" ");
            write!(f, "{}. {}", vars_str, self.ty)
        }
    }
}

// --- Helper methods for finding free type variables ---
impl Type {
    pub fn free_type_vars(&self) -> HashSet<TypeVar> {
        let mut ftv = HashSet::new();
        self.collect_ftv(&mut ftv);
        ftv
    }

    fn collect_ftv(&self, ftv: &mut HashSet<TypeVar>) {
        match self {
            Type::Var(tv) => { ftv.insert(tv.clone()); }
            Type::Fun(t1, t2) => {
                t1.collect_ftv(ftv);
                t2.collect_ftv(ftv);
            }
            Type::List(t) => t.collect_ftv(ftv),
            _ => {}
        }
    }
}
