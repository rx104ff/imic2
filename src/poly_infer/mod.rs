pub mod ast;
pub mod parser;
pub mod proof;
pub mod token;
pub mod poly_infer;
pub mod unifier;