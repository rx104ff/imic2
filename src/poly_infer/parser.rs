use std::collections::{HashMap, HashSet};

use crate::common::ast::{Type, PolyTypeEnv, TyScheme, TypeVar, Judgment};
use crate::common::parser::{ExpressionParser, ParserCore, TypeParser};
use crate::common::tokenizer::Token;


/// A recursive descent parser for the TypingML4 language.
pub struct Parser {
    core: ParserCore,
    type_var_map: HashMap<String, TypeVar>,
    next_parser_var_id: usize,
}

impl ExpressionParser for Parser {
    fn core(&mut self) -> &mut ParserCore {
        &mut self.core
    }
}

impl TypeParser for Parser {
    fn core(&mut self) -> &mut ParserCore {
        &mut self.core
    }

    fn parse_single_type(&mut self) -> Result<Type, String> {
        let ty = match self.core.peek().cloned() {
            Some(Token::Ident(name)) => {
                self.core.advance();
                match name.as_str() {
                    "int" => Type::Int,
                    "bool" => Type::Bool,
                    _ => return Err(format!("Unknown type name '{}'", name)),
                }
            }
            Some(Token::TypeVar(name)) => {
                self.core.advance();
                let tv = self.get_or_create_parser_var(name);
                Type::Var(tv)
            }
            Some(Token::LParen) => {
                self.core.advance();
                let inner_ty = self.parse_type()?;
                self.core.expect(Token::RParen)?;
                inner_ty
            }
            _ => return Err("Expected a type name or a parenthesized type.".to_string()),
        };
        if let Some(Token::Ident(s)) = self.core.peek() {
            if s == "list" {
                self.core.advance();
                return Ok(Type::List(Box::new(ty)));
            }
        }
        Ok(ty)
    }
}

impl Parser {
    pub fn new(tokens: Vec<Token>) -> Self {
        Self {
            core: ParserCore::new(tokens),
            type_var_map: HashMap::new(),
            next_parser_var_id: 0,
        }
    }

    /// The main entry point for the parser.
    /// It parses a judgment of the form `env |- expr : type`
    /// and returns the parsed Judgment struct.
    pub fn parse(&mut self) -> Result<(Judgment, HashSet<String>), String> {
        let env = self.parse_type_env()?;
        self.core.expect(Token::Turnstile)?;
        let expr = self.parse_expr()?;
        self.core.expect(Token::Colon)?;
        let ty = self.parse_type()?;
        let used_names = self.type_var_map.keys().map(|s| format!("'{}", s)).collect();
        Ok((Judgment::PolyInfer(env, expr, ty), used_names))
    }

    fn get_or_create_parser_var(&mut self, name: String) -> TypeVar {
        if let Some(var) = self.type_var_map.get(&name) {
            return var.clone();
        }
        let id = self.next_parser_var_id;
        self.next_parser_var_id += 1;
        let tv = TypeVar { id, name: format!("'{}", name) };
        self.type_var_map.insert(name, tv.clone());
        tv
    }

    // --- Type Environment and Type Parsing ---
    fn parse_type_env(&mut self) -> Result<PolyTypeEnv, String> {
        let mut env = PolyTypeEnv::new();
        if self.core.peek() == Some(&Token::Turnstile) { return Ok(env); }
        loop {
            let var = self.core.next_var()?;
            self.core.expect(Token::Colon)?;
            let scheme = self.parse_type_scheme()?;
            env.push((var, scheme));
            if self.core.peek() == Some(&Token::Comma) { self.core.advance(); } 
            else { break; }
        }
        Ok(env)
    }

    // Parses a full type scheme, including `forall` quantifiers.
    fn parse_type_scheme(&mut self) -> Result<TyScheme, String> {
        let mut quantified_vars = vec![];
        let mut potential_var_names = vec![];
        let initial_pos = self.core.pos();

        while let Some(Token::TypeVar(name)) = self.core.peek().cloned() {
            potential_var_names.push(name);
            self.core.advance();
        }

        if self.core.peek() == Some(&Token::Dot) {
            self.core.advance();
            potential_var_names.sort();
            for name in potential_var_names {
                let tv = self.get_or_create_parser_var(name);
                quantified_vars.push(tv);
            }
        } else {
            self.core.initial(initial_pos);
        }

        let ty = self.parse_type()?;
        Ok(TyScheme { vars: quantified_vars, ty })
    }    
}
