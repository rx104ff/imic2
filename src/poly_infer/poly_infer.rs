use std::collections::{BTreeMap, HashSet};
use crate::common::ast::{Expr, Op, Type, TypeVar, TyScheme, TypeEnv, Judgment};
use crate::poly_infer::proof::Derivation;
use crate::poly_infer::unifier::{unify, apply_sub, Substitution};

// The context for inference, holding substitutions.
struct InferContext {
    sub: Substitution,
    var_counter: usize,
    used_names: HashSet<String>,
}

impl InferContext {
    fn new_type_var(&mut self) -> Type {
        let mut name_id = 0;
        loop {
            let name = format!("'{}", ((name_id % 26) as u8 + b'a') as char);
            if !self.used_names.contains(&name) {
                self.used_names.insert(name.clone());
                let id = self.var_counter;
                self.var_counter += 1;
                // Use a large offset to ensure inferrer IDs do not collide with parser IDs.
                print!("{}", name);
                return Type::Var(TypeVar { id: id + 1000, name });
            }
            name_id += 1;
        }
    }
}

/// The main public entry point for the type inferrer.
/// It takes a judgment `Γ ⊢ e`, infers its type, and returns a full derivation proof.
pub fn infer_judgment(judgment: &Judgment, used_names: HashSet<String>) -> Result<Derivation, String> {
    match judgment {
        // Handle the case for polymorphic type inference (PolyTypingML4)
        Judgment::PolyInfer(env, expr, expected_ty) => {
            let mut ctx = InferContext {
                sub: BTreeMap::new(),
                var_counter: 0,
                used_names
            };
            // We infer the expression's type, with polymorphism enabled.
            let mut inferred_derivation = infer_expr(&mut ctx, env, expr)?;
            
            let final_sub = unify(&inferred_derivation.ty, expected_ty, &ctx.sub)?;
            ctx.sub = final_sub;

            apply_sub_to_deriv(&mut inferred_derivation, &ctx.sub);
            inferred_derivation.ty = apply_sub(expected_ty, &ctx.sub);
            Ok(inferred_derivation)
        }
        _ => Err("This judgment type is not supported by the type checker.".to_string()),
    }
}

/// The recursive helper that generates and solves type constraints,
/// while simultaneously building the derivation tree.
fn infer_expr(ctx: &mut InferContext, env: &TypeEnv, e: &Expr) -> Result<Derivation, String> {
    match e {
        Expr::Int(_) => Ok(Derivation {
            env: env.clone(), expr: e.clone(), ty: Type::Int,
            rule: "T-Int".to_string(), premises: vec![],
        }),
        Expr::Bool(_) => Ok(Derivation {
            env: env.clone(), expr: e.clone(), ty: Type::Bool,
            rule: "T-Bool".to_string(), premises: vec![],
        }),
        Expr::Nil => Ok(Derivation {
            env: env.clone(), expr: e.clone(), ty: Type::List(Box::new(ctx.new_type_var())),
            rule: "T-Nil".to_string(), premises: vec![],
        }),
        Expr::Var(var) => {
            for (v, scheme) in env.iter().rev() {
                if v == var {
                    // When using a variable, we instantiate its type scheme
                    // to get a fresh type with new type variables.
                    let ty = instantiate(scheme, ctx);
                    return Ok(Derivation {
                        env: env.clone(), expr: e.clone(), ty,
                        rule: "T-Var".to_string(), premises: vec![],
                    });
                }
            }
            Err(format!("Unbound variable: {}", var.0))
        }
        Expr::Fun(param, body, _) => {
            let param_ty = ctx.new_type_var();
            let mut new_env = env.clone();
            // In a `fun`, the parameter is monomorphic (not a `forall` type).
            new_env.push((param.clone(), TyScheme { vars: vec![], ty: param_ty.clone() }));
            
            let body_deriv = infer_expr(ctx, &new_env, body)?;
            
            let fun_ty = Type::Fun(Box::new(apply_sub(&param_ty, &ctx.sub)), Box::new(body_deriv.ty.clone()));
            
            Ok(Derivation {
                env: env.clone(), expr: e.clone(), ty: fun_ty,
                rule: "T-Abs".to_string(), premises: vec![body_deriv],
            })
        }
        Expr::App(e1, e2, _) => {
            let d1 = infer_expr(ctx, env, e1)?;
            let t1 = apply_sub(&d1.ty, &ctx.sub);
            let d2 = infer_expr(ctx, env, e2)?;
            let t2 = apply_sub(&d2.ty, &ctx.sub);

            let return_ty = ctx.new_type_var();
            let fun_ty = Type::Fun(Box::new(t2), Box::new(return_ty.clone()));
            
            ctx.sub = unify(&t1, &fun_ty, &ctx.sub)?;
            
            Ok(Derivation {
                env: env.clone(), expr: e.clone(), ty: return_ty,
                rule: "T-App".to_string(), premises: vec![d1, d2],
            })
        }
        Expr::Let(x, e1, e2, _) => {
            let d1 = infer_expr(ctx, env, e1)?;
            let t1 = apply_sub(&d1.ty, &ctx.sub);

            let mut new_env = env.clone();
            // (GEN) rule: generalize the type of the bound expression.
            let scheme = generalize(env, &t1, &ctx.sub);
            new_env.push((x.clone(), scheme));
            
            let d2 = infer_expr(ctx, &new_env, e2)?;
            
            Ok(Derivation {
                env: env.clone(), expr: e.clone(), ty: d2.ty.clone(),
                rule: "T-Let".to_string(), premises: vec![d1, d2],
            })
        }
        Expr::If(cond, then_branch, else_branch, _) => {
            let d_cond = infer_expr(ctx, env, cond)?;
            ctx.sub = unify(&d_cond.ty, &Type::Bool, &ctx.sub)?;
            let d_then = infer_expr(ctx, env, then_branch)?;
            let d_else = infer_expr(ctx, env, else_branch)?;
            ctx.sub = unify(&d_then.ty, &d_else.ty, &ctx.sub)?;
            Ok(Derivation {
                env: env.clone(), expr: e.clone(), ty: d_then.ty.clone(),
                rule: "T-If".to_string(), premises: vec![d_cond, d_then, d_else],
            })
        }
        Expr::LetRec(f, x, e1, e2, _) => {
            let t1 = ctx.new_type_var();
            let t2 = ctx.new_type_var();
            let fun_ty = Type::Fun(Box::new(t1.clone()), Box::new(t2.clone()));
            let mut new_env1 = env.clone();
            new_env1.push((f.clone(), TyScheme { vars: vec![], ty: fun_ty.clone() }));
            new_env1.push((x.clone(), TyScheme { vars: vec![], ty: t1 }));
            let d1 = infer_expr(ctx, &new_env1, e1)?;
            ctx.sub = unify(&d1.ty, &t2, &ctx.sub)?;
            let mut new_env2 = env.clone();
            new_env2.push((f.clone(), generalize(env, &fun_ty, &ctx.sub)));
            let d2 = infer_expr(ctx, &new_env2, e2)?;
            Ok(Derivation {
                env: env.clone(), expr: e.clone(), ty: d2.ty.clone(),
                rule: "T-LetRec".to_string(), premises: vec![d1, d2],
            })
        }
        Expr::BinOp(e1, op, e2, _) => {
            let d1 = infer_expr(ctx, env, e1)?;
            let t1 = apply_sub(&d1.ty, &ctx.sub);
            let d2 = infer_expr(ctx, env, e2)?;
            let t2 = apply_sub(&d2.ty, &ctx.sub);
            
            let (expected_t1, expected_t2, result_ty, rule_name) = match op {
                Op::Add => (Type::Int, Type::Int, Type::Int, "T-Plus"),
                Op::Sub => (Type::Int, Type::Int, Type::Int, "T-Minus"),
                Op::Mul => (Type::Int, Type::Int, Type::Int, "T-Mult"),
                Op::Lt => (Type::Int, Type::Int, Type::Bool, "T-Lt"),
                Op::Cons => {
                    let elem_type = ctx.new_type_var();
                    (elem_type.clone(), Type::List(Box::new(elem_type)), t2.clone(), "T-Cons")
                }
            };
            
            let sub1 = unify(&t1, &expected_t1, &ctx.sub)?;
            let sub2 = unify(&t2, &expected_t2, &sub1)?;
            ctx.sub = sub2;
            
            let final_type = apply_sub(&result_ty, &ctx.sub);
            Ok(Derivation {
                env: env.clone(),
                expr: e.clone(),
                ty: final_type,
                rule: rule_name.to_string(),
                premises: vec![d1, d2]
            })
        }
        Expr::Match(e1, e2, x, y, e3, _) => {
            let d1 = infer_expr(ctx, env, e1)?;
            let t1 = apply_sub(&d1.ty, &ctx.sub);
            let elem_ty = ctx.new_type_var();
            let list_ty = Type::List(Box::new(elem_ty.clone()));
            ctx.sub = unify(&t1, &list_ty, &ctx.sub)?;
            let d2 = infer_expr(ctx, env, e2)?;
            let t_nil = apply_sub(&d2.ty, &ctx.sub);
            let mut new_env = env.clone();
            new_env.push((x.clone(), TyScheme { vars: vec![], ty: elem_ty }));
            new_env.push((y.clone(), TyScheme { vars: vec![], ty: list_ty }));
            let d3 = infer_expr(ctx, &new_env, e3)?;
            let t_cons = apply_sub(&d3.ty, &ctx.sub);
            ctx.sub = unify(&t_nil, &t_cons, &ctx.sub)?;
            Ok(Derivation {
                env: env.clone(), expr: e.clone(), ty: t_nil,
                rule: "T-Match".to_string(), premises: vec![d1, d2, d3],
            })
        }
        _ => Err("This judgment type is not supported by the type checker.".to_string()),
    }
}

// --- Polymorphism and Finalization Helpers ---
fn generalize(env: &TypeEnv, ty: &Type, sub: &Substitution) -> TyScheme {
    let ty = apply_sub(ty, sub);
    let mut env_ftv = HashSet::new();
    for (_, scheme) in env {
        env_ftv.extend(apply_sub(&scheme.ty, sub).free_type_vars());
    }
    let ty_ftv = ty.free_type_vars();
    
    // Collect the difference into a Vec
    let mut quantified_vars: Vec<_> = ty_ftv.difference(&env_ftv).cloned().collect();
    
    // Sort the quantified variables by their ID to ensure a canonical order.
    // This is crucial for deterministic instantiation and display.
    quantified_vars.sort_by_key(|v| v.id);
    
    TyScheme { vars: quantified_vars, ty }
}

fn instantiate(scheme: &TyScheme, ctx: &mut InferContext) -> Type {
    let mut fresh_sub = Substitution::new();
    for var in &scheme.vars {
        fresh_sub.insert(var.clone(), ctx.new_type_var());
    }
    apply_sub(&scheme.ty, &fresh_sub)
}

fn apply_sub_to_env(env: &TypeEnv, sub: &Substitution) -> TypeEnv {
    env.iter()
        .map(|(var, scheme)| {
            // We must not substitute the variables that are quantified by this scheme.
            let mut temp_sub = sub.clone();
            for quantified_var in &scheme.vars {
                temp_sub.remove(quantified_var);
            }
            let new_ty = apply_sub(&scheme.ty, &temp_sub);
            (var.clone(), TyScheme { vars: scheme.vars.clone(), ty: new_ty })
        })
        .collect()
}

/// Applies substitutions to all types within a derivation tree.
fn apply_sub_to_deriv(deriv: &mut Derivation, sub: &Substitution) {
    deriv.ty = apply_sub(&deriv.ty, sub);
    deriv.env = apply_sub_to_env(&deriv.env, sub); // Use the corrected helper
    for premise in &mut deriv.premises {
        apply_sub_to_deriv(premise, sub);
    }
}
