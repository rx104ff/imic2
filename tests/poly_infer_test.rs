use imic2::poly_infer;
use insta::assert_snapshot;

/// A helper function to simulate a full run of the polymorphic type inferrer.
/// It tokenizes, parses, and infers the type for a given judgment string.
fn run_poly_test(input: &str) -> String {
    poly_infer::ast::reset_type_var_counter();
    let tokens = poly_infer::token::tokenize(input);
    let mut parser = poly_infer::parser::Parser::new(tokens);
    let (judgment, used_names) = match parser.parse_judgment() {
        Ok(j) => j,
        Err(e) => return format!("Parsing Error: {}", e),
    };
    match poly_infer::poly_infer::infer_judgment(&judgment, used_names) {
        Ok(derivation) => {
            // The derivation object contains the full proof tree with all types resolved.
            // We can now print it directly.
            format!("{}", derivation)
        }
        Err(e) => format!("Type Error: {}", e),
    }
}

#[test]
fn test_infer_poly_typing_107() {
    let result = run_poly_test("|- fun x -> x : 'a -> 'a");
    assert_snapshot!(&result);
}

#[test]
fn test_infer_poly_typing_108() {
    let result = run_poly_test("f: 'a.'a->'a |- f 3 : int");
    assert_snapshot!(&result);
}

#[test]
fn test_infer_poly_typing_109() {
    let result = run_poly_test("f: 'a.'a->'a |- f (fun x -> x + 3) : int -> int");
    assert_snapshot!(&result);
}

#[test]
fn test_infer_poly_typing_110() {
    let result = run_poly_test("|- let id = fun x -> x in id id : bool -> bool");
    assert_snapshot!(&result);
}

#[test]
fn test_infer_poly_typing_111() {
    let result = run_poly_test("f: 'a 'b.'a->'b->'a |- f 3 true + f 2 4 : int");
    assert_snapshot!(&result);
}

#[test]
fn test_infer_poly_typing_112() {
    let result = run_poly_test("|- let k = fun x -> fun y -> x in (k 3 true) :: (k (1::[]) 3) : int list");
    assert_snapshot!(&result);
}

#[test]
fn test_infer_poly_typing_113() {
    let result = run_poly_test("|- let compose = fun f -> fun g -> fun x -> f (g x) in
   let f = fun x -> if x then 3 else 4 in
   let g = fun x -> x < 4 in
   compose f (compose g f) true : int");
    assert_snapshot!(&result);
}

#[test]
fn test_infer_poly_typing_114() {
    let result = run_poly_test("|- let twice = fun f -> fun x -> f (f x) in
   twice (fun x -> x + 4) 5 : int");
    assert_snapshot!(&result);
}

#[test]
fn test_infer_poly_typing_115() {
    let result = run_poly_test("|- let twice = fun f -> fun x -> f (f x) in
   twice twice (fun x -> x + 4) 5 : int");
    assert_snapshot!(&result);
}

// TODO: This test produces inconsistant output
#[test]
fn test_infer_poly_typing_116() {
    let result = run_poly_test("|- let s = fun f -> fun g -> fun x -> f x (g x) in
   let k = fun x -> fun y -> x in
   s k k : 'a -> 'a");
    assert_snapshot!(&result);
}

#[test]
fn test_infer_poly_typing_117() {
    let result = run_poly_test("|- let x = [] in let y = 3 :: x in true :: x : bool list");
    assert_snapshot!(&result);
}

#[test]
fn test_infer_poly_typing_118() {
    
    let result = run_poly_test("|- let l = (fun x -> x) :: [] in
   let l1 = (fun y -> y + 1) :: l in
   (fun z -> if z then false else true) :: l : (bool -> bool) list");
    assert_snapshot!(&result);
}
