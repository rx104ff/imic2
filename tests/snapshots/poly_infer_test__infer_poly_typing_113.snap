---
source: tests/poly_infer_test.rs
expression: "&result"
---
|- let compose = fun f -> fun g -> fun x -> f (g x) in let f = fun x -> if x then 3 else 4 in let g = fun x -> x < 4 in compose f (compose g f) true : int by T-Let {
    |- fun f -> fun g -> fun x -> f (g x) : (('d -> 'e) -> (('c -> 'd) -> ('c -> 'e))) by T-Abs {
        f: ('d -> 'e) |- fun g -> fun x -> f (g x) : (('c -> 'd) -> ('c -> 'e)) by T-Abs {
            f: ('d -> 'e), g: ('c -> 'd) |- fun x -> f (g x) : ('c -> 'e) by T-Abs {
                f: ('d -> 'e), g: ('c -> 'd), x: 'c |- f (g x) : 'e by T-App {
                    f: ('d -> 'e), g: ('c -> 'd), x: 'c |- f : ('d -> 'e) by T-Var {
                    };
                    f: ('d -> 'e), g: ('c -> 'd), x: 'c |- (g x) : 'd by T-App {
                        f: ('d -> 'e), g: ('c -> 'd), x: 'c |- g : ('c -> 'd) by T-Var {
                        };
                        f: ('d -> 'e), g: ('c -> 'd), x: 'c |- x : 'c by T-Var {
                        };
                    };
                };
            };
        };
    };
    compose: 'c 'd 'e. (('d -> 'e) -> (('c -> 'd) -> ('c -> 'e))) |- let f = fun x -> if x then 3 else 4 in let g = fun x -> x < 4 in compose f (compose g f) true : int by T-Let {
        compose: 'c 'd 'e. (('d -> 'e) -> (('c -> 'd) -> ('c -> 'e))) |- fun x -> if x then 3 else 4 : (bool -> int) by T-Abs {
            compose: 'c 'd 'e. (('d -> 'e) -> (('c -> 'd) -> ('c -> 'e))), x: bool |- if x then 3 else 4 : int by T-If {
                compose: 'c 'd 'e. (('d -> 'e) -> (('c -> 'd) -> ('c -> 'e))), x: bool |- x : bool by T-Var {
                };
                compose: 'c 'd 'e. (('d -> 'e) -> (('c -> 'd) -> ('c -> 'e))), x: bool |- 3 : int by T-Int {
                };
                compose: 'c 'd 'e. (('d -> 'e) -> (('c -> 'd) -> ('c -> 'e))), x: bool |- 4 : int by T-Int {
                };
            };
        };
        compose: 'c 'd 'e. (('d -> 'e) -> (('c -> 'd) -> ('c -> 'e))), f: (bool -> int) |- let g = fun x -> x < 4 in compose f (compose g f) true : int by T-Let {
            compose: 'c 'd 'e. (('d -> 'e) -> (('c -> 'd) -> ('c -> 'e))), f: (bool -> int) |- fun x -> x < 4 : (int -> bool) by T-Abs {
                compose: 'c 'd 'e. (('d -> 'e) -> (('c -> 'd) -> ('c -> 'e))), f: (bool -> int), x: int |- x < 4 : bool by T-Lt {
                    compose: 'c 'd 'e. (('d -> 'e) -> (('c -> 'd) -> ('c -> 'e))), f: (bool -> int), x: int |- x : int by T-Var {
                    };
                    compose: 'c 'd 'e. (('d -> 'e) -> (('c -> 'd) -> ('c -> 'e))), f: (bool -> int), x: int |- 4 : int by T-Int {
                    };
                };
            };
            compose: 'c 'd 'e. (('d -> 'e) -> (('c -> 'd) -> ('c -> 'e))), f: (bool -> int), g: (int -> bool) |- compose f (compose g f) true : int by T-App {
                compose: 'c 'd 'e. (('d -> 'e) -> (('c -> 'd) -> ('c -> 'e))), f: (bool -> int), g: (int -> bool) |- compose f (compose g f) : (bool -> int) by T-App {
                    compose: 'c 'd 'e. (('d -> 'e) -> (('c -> 'd) -> ('c -> 'e))), f: (bool -> int), g: (int -> bool) |- compose f : ((bool -> bool) -> (bool -> int)) by T-App {
                        compose: 'c 'd 'e. (('d -> 'e) -> (('c -> 'd) -> ('c -> 'e))), f: (bool -> int), g: (int -> bool) |- compose : ((bool -> int) -> ((bool -> bool) -> (bool -> int))) by T-Var {
                        };
                        compose: 'c 'd 'e. (('d -> 'e) -> (('c -> 'd) -> ('c -> 'e))), f: (bool -> int), g: (int -> bool) |- f : (bool -> int) by T-Var {
                        };
                    };
                    compose: 'c 'd 'e. (('d -> 'e) -> (('c -> 'd) -> ('c -> 'e))), f: (bool -> int), g: (int -> bool) |- (compose g f) : (bool -> bool) by T-App {
                        compose: 'c 'd 'e. (('d -> 'e) -> (('c -> 'd) -> ('c -> 'e))), f: (bool -> int), g: (int -> bool) |- compose g : ((bool -> int) -> (bool -> bool)) by T-App {
                            compose: 'c 'd 'e. (('d -> 'e) -> (('c -> 'd) -> ('c -> 'e))), f: (bool -> int), g: (int -> bool) |- compose : ((int -> bool) -> ((bool -> int) -> (bool -> bool))) by T-Var {
                            };
                            compose: 'c 'd 'e. (('d -> 'e) -> (('c -> 'd) -> ('c -> 'e))), f: (bool -> int), g: (int -> bool) |- g : (int -> bool) by T-Var {
                            };
                        };
                        compose: 'c 'd 'e. (('d -> 'e) -> (('c -> 'd) -> ('c -> 'e))), f: (bool -> int), g: (int -> bool) |- f : (bool -> int) by T-Var {
                        };
                    };
                };
                compose: 'c 'd 'e. (('d -> 'e) -> (('c -> 'd) -> ('c -> 'e))), f: (bool -> int), g: (int -> bool) |- true : bool by T-Bool {
                };
            };
        };
    };
};
